_labs/lab08.markdown

```diff
@@ -0,0 +1,171 @@
+The `main` function of `src/Constraint.cpp` ties this logic together.
+
+In short, the lab consists of the following tasks:
+
+1.) Write Datalog rules in the initialize function in `Extractor.cpp` to define the reaching definition analysis and taint analysis.
+2.) Write the `extractContraints` function in `Extractor.cpp` that extracts Datalog facts from LLVM IR `Instruction`.
+
+**Relations for Datalog Analysis**. The skeleton code provides the definitions of necessary Datalog relations over LLVM IR in `Extractor.h`.
+In the following subsection, we will show how to represent LLVM IR programs using these relations.
+
+The relations for def and use of variables are as follows:
+
+- `def(var,inst)`: Variable var is defined at instruction inst.
+- `use(var,inst)`: Variable var is used at instruction inst.
+
+The relations for the reaching definition analysis are as follows: 
+
+- `kill(curr_inst, old_inst)`: Instruction curr_inst kills definition at instruction old_inst.
+- `next(curr_inst, next_inst)`: Instruction curr_inst is an immediate successor of instruction next_inst.
+- `in(inst, def_inst)`: Definition at defining instruction def_inst may reach the program point just before instruction inst.
+- `out(inst, def_inst)`: Definition at defining instruction def_inst may reach the program point just after instruction inst.
+
+Note that the `Kill` relation can be derived by using the `Def` relation by writing a Datalog rule.
+
+The relations for the taint analysis are as follows: 
+
+- `taint(inst)` : There exists a function call at intruction inst that reads a tainted input.
+- `edge(from, to)`: There exists an immediate data-flow from instruction `from` to instruction `to`.
+- `Path(from, to)`: There exists a transitive tainted data-flow from instruction `from` to instruction `to`.
+- `sanitizer(inst)` : There exists a function call at intruction inst that sanitizes a tainted input.
+- `div(denom, inst)` : There exists a division operation at instruction inst whose divisor is variable denom.
+- `alarm(inst)` : There exists a potential exploitable divide-by-zero error at instruction inst.
+
+
+Assume that input programs may contain function calls to `tainted_input` and `sanitizer` that read and sanitize a tainted input, respectively. The final output relation for potential bug reports is `alarm`.
+
+You will use these relations to build rules for the definition analysis and taint analysis in `Extractor.cpp`.
+
+**Encoding LLVM Instruction in Datalog**. Recall that, in LLVM IR, values and instructions are interchangable. Therefore, all variables X, Y, and Z are an instance of LLVM’s `Value` class. 
+
+Assume that input C programs do not have pointer variables. Therefore, we abuse pointer variables in LLVM IR as their dereference expressions. Consider the following simplified LLVM program from a simple C program `int x = 0; int y = x;`:
+
+```sh
```

Comment:
This is LLVM IR code not shell script, see prev labs to see the correct annotation for llvm code highlight, its likely `llvm`